### PR TITLE
fix Issue 22631 - ImportC: support C++11 unscoped enums with underlyi…

### PR DESCRIPTION
**Pull request by WalterBright**
…ng type

Easier to implement it than argue about it being an extension.

### CODE REVIEW

**Review by ibuclaw — CHANGES_REQUESTED** [review]
I'd prefer to have this verified first. Has the test been confirmed to be accurate?

**Comment by WalterBright** [comment]
Have what verified? what test?

**Comment by ibuclaw** [comment]
> Have what verified? what test?

For instance, is it clang in "C" mode that's being ran?  Don't want to add C++ features just because the test conditions are wrong (GCC C rejects this code, and rightly so).

**Comment by WalterBright** [comment]
My version of clang won't accept it, but it's old.

**Comment by ryuukk** [comment]
Wow that is very interesting!

I always dreamed to get that feature in D, it makes writing flags so much nicer, and overall playing with enums when the type name is explicitly long, less typing!

Is there a chance to get that feature in D? rather than just ImportC?

File: src/dmd/dsymbolsem.d [hunk 1]
```diff
@@ -2179,7 +2179,7 @@ private extern(C++) final class DsymbolSemanticVisitor : Visitor
         {
             /* C11 6.7.2.2
              */
-            ed.memtype = Type.tint32; // C11 6.7.2.2-4 implementation defined
+            assert(ed.memtype);
             int nextValue = 0;        // C11 6.7.2.2-3 first member value defaults to 0
```
Comment (moon-chilled) [inline src/dmd/dsymbolsem.d:2183]: Couldn't this overflow if the base type is specified as e.g. long?

File: src/dmd/cparse.d [hunk 2]
```diff
@@ -3075,6 +3075,17 @@ final class CParser(AST) : Parser!AST
             nextToken();
         }
 
+        /* Extension: add optional base type after the identifier
```
Comment (ibuclaw) [inline src/dmd/cparse.d:3078]:
```suggestion
        /* Clang extension: add optional base type after the identifier
```

**Review by ibuclaw — COMMENTED** [review]
OK, I can see it.
```
warning: enumeration types with a fixed underlying type are a Clang extension [-Wfixed-enum-extension]
enum E : char { A = 3, B };
       ^~~~~~
```

File: src/dmd/cparse.d (hunk 2 above)
Comment (moon-chilled) [reply src/dmd/cparse.d:3078]: I think this was proposed for c2x as well, not sure if it made it in though

Comment (ibuclaw) [reply src/dmd/cparse.d:3078]: It's not even mentioned in the current working draft.

**Comment by ibuclaw** [comment]
> Wow that is very interesting!
> 
> I always dreamed to get that feature in D, it makes writing flags so much nicer, and overall playing with enums when the type name is explicitly long, less typing!
> 
> Is there a chance to get that feature in D? rather than just ImportC?

Eh?  It's always been available in D.

File: src/dmd/dsymbolsem.d (hunk 1 above)
Comment (WalterBright) [reply src/dmd/dsymbolsem.d:2183]: No more than it would in D.

File: src/dmd/cparse.d (hunk 2 above)
Comment (WalterBright) [reply src/dmd/cparse.d]: done

**Comment by WalterBright** [comment]
@ibuclaw ready to rock!

**Comment by ryuukk** [comment]
> > Wow that is very interesting!
> > I always dreamed to get that feature in D, it makes writing flags so much nicer, and overall playing with enums when the type name is explicitly long, less typing!
> > Is there a chance to get that feature in D? rather than just ImportC?
> 
> Eh? It's always been available in D.

@ibuclaw 

I edited my comment to be more precise
Allowing this:

```D
enum Flag { A = 1, B = 2, C = 4 }
Flag flags = A | B | C;
```

**Review by ibuclaw — APPROVED** [review]

**Comment by ibuclaw** [comment]
> I edited my comment to be more precise Allowing this:
> 
> ```d
> enum Flag { A = 1, B = 2, C = 4 }
> Flag flags = A | B | C;
> ```

`enum Flag` becomes `enum : Flag`, and `Flag` is an `alias` to whatever underlying type you want.

**Comment by ryuukk** [comment]
> ```d
> enum Flag { A = 1, B = 2, C = 4 }
> Flag flags = A | B | C;
> ```

wow i didn't know that was possible! thanks a lot!!

**Comment by maxhaton** [comment]
What C libraries require this to be imported?

**Pull request merged**